docs/block-processing/types.md

```diff
@@ -0,0 +1,1152 @@
+
+  * `_gbhPrevBlock :: HeaderHash` – a hash of the previous block's header.
+
+  * `_gbhBodyProof :: BodyProof MainBlockchain` – proofs (e.g. hashes) of
+    payloads from the block body; they will be discussed in the section
+    about [block payload](#main-block-payload-proof).
+
+  * `_gbhConsensus :: ConsensusData MainBlockchain` – meta-information about
+    the block:
+
+      + `_mcdSlot :: SlotId` – the slot for which the block was generated
+      + `_mcdLeaderKey :: PublicKey` – public key of the slot leader (which
+        may be different from the block issuer, because of delegation)
+      + `_mcdDifficulty :: ChainDifficulty` – difficulty of the chain ending
+        in this block (i.e. number of main blocks between the first block
+        ever and this block, inclusive)
+      + `_mcdSignature :: BlockSignature` – a signature of the block by its
+        issuer (if the issuer isn't the block leader, it will be a delegated
+        signature confirming issuer's right to issue the block in this slot)
+
+  * `_gbhExtra :: MainExtraHeaderData` – more information about the block:
+
+      + `_mehBlockVersion :: BlockVersion` – the block version;
+        see [Software and block versions](us.md#software-and-block-versions)
+      + `_mehSoftwareVersion :: SoftwareVersion` – ditto
+      + `_mehEBDataProof :: Hash MainExtraBodyData` – a hash of the extra
+        data in the block (since ultimately a header needs to checksum *all*
+        data in the block)
+      + `_mehAttributes :: BlockHeaderAttributes` – an attributes map to
+        extend the header with more fields, currently empty
+
+Invariants:
+
+  * The signature must be valid.
+
+  * If the signature is a delegated signature (`BlockPSignatureLight` or
+    `BlockPSignatureHeavy`), it must not be self-signed – i.e. the
+    `pskIssuerPk` and `pskDelegatePk` must be different.
+
+### Main block signature
+
+*keywords: `BlockSignature`, `MainToSign`*
+
+A `BlockSignature` verifies that the block was issued by someone who had a
+right to issue it, and also verifies that the block hasn't been tampered
+with:
+
+```haskell
+data BlockSignature
+    = BlockSignature (Signature MainToSign)
+    | BlockPSignatureLight (ProxySigLight MainToSign)
+    | BlockPSignatureHeavy (ProxySigHeavy MainToSign)
+```
+
+There are three kinds of signatures:
+
+  * `BlockSignature` – simply a signature of the block by its issuer
+
+  * `BlockPSignatureLight` – a light delegation signature (not discussed in
+    this document); this signature is used when the right to issue blocks
+    has been transferred on a temporary basis
+
+  * `BlockPSignatureHeavy` – a heavy delegation signature, used when the
+    right to issue blocks has been transferred “until further notice”
+
+The thing signed by a `BlockSignature` isn't the whole block, but only
+certain parts of it (this is done because signing is an expensive process
+and we'd like to sign as few bytes as we can get away with). The exact parts
```

Comment:
The way it's written now is ok, but I just want to say that we sign basically `MainHeader \ BlockSignature`. The reason why we sign `MainHeader` (and not block) is because we'd like to sign fewer bytes (as you said) and the reason why we exclude `BlockSignature` is that it'd be impossible otherwise :)